File: Data Structures & Algorithms/binary-search/submission-1.go
func search(nums []int, target int) int {
    var left, right int = 0, len(nums)-1

    for left <= right {
        // calculate center
        center := left + (right - left) / 2
        // if target is larger, then left = center
        if target > nums[center] {
            left = center + 1
        } else if target < nums[center] {
            right = center - 1
        } else {
            return center
        }
        // if target is smaller, then right = center
        // if center == target return center
    }

    return -1

    // O(logn)
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # binary search
        # before we do the regular boundary move, check:
        # if right value is less than center value:
        # right = center
        # if left value is greater than center value:
        # left = center

        left, right = 0, len(nums) - 1

        while left <= right:
            center = left + (right - left)//2
            
            if nums[right] < nums[center]:
                left = center + 1
            elif nums[right] > nums[center]:
                right = center
            else:
                break

        return nums[left]
        
        